### ISSUE TITLE
Add spellchecking

### ISSUE THREAD

**Issue by sigod** [telegramdesktop/tdesktop#96]

**Comment by Innomen** [comment]
Still wanted. (English.) Notepad++ has a spell check plugin, perhaps someone could adapt that?

**Comment by zumoshi** [comment]
would be nice to have.

**Comment by HallEffect** [comment]
It will be nice feature!

**Comment by alleyoopster** [comment]
Pitching in here to plead for this to get implemented. My smelling needs checking :)

**Comment by Innomen** [comment]
"My smelling needs checking :)"

I lold4r :)

**Comment by WyRe** [comment]
+1

**Comment by chaosdrop** [comment]
+1

**Comment by maronin** [comment]
+1

**Comment by rkhitin** [comment]
+1

**Comment by Lev91gor** [comment]
+2.71

**Comment by Sushubh** [comment]
+1. Too used to Chrome apps which have access to the brilliant Google's own spell checker powered by the web. A basic one would be much appreciated. Thanks.

**Comment by Zignd** [comment]
+1 I'd really appreciate if you guys could add this feature to the application.

**Comment by chaosdrop** [comment]
Just figured I'd post it here since it's what I am doing in the mean time.

**Comment by davidsteltz** [comment]
+1
+99,999,999 really

**Comment by NeonMonk** [comment]
C'mon guys. This can't be that hard!

**Comment by gloeglm** [comment]
It really can't be ... all you need is some kind of party that you trust you can exchange keys through ... And the native mac client already does that?

**Comment by chaosdrop** [comment]
In most open source projects it's all about integrating Enchant (since it works on OSX, Windows, Linux and can cover a huge number of languages) unless they want to do some kind of special in-house spell-checker. Which I assume most be what they are looking into having taken this long, normally Enchant is pretty fast and easy to get worked into a project.

**Comment by DmytroLukianchuk** [comment]
+100500 really looking forward to seeing this feature would be done for this amazing app! 
Would it be ever? 
have 0.9.2 version.

**Comment by comzeradd** [comment]
On Linux the best solution would be to utilize hunspell for this.

**Comment by jacobmischka** [comment]
So many people shaming the devs and so zero people submitting pull requests.

**Comment by Innomen** [comment]
@jacobmischka 
You know I hear whtieknight shield polish is still on sale from Cyber Monday in a few places. >.>

**Comment by auchri** [comment]
@jacobmischka +1

**Comment by Innomen** [comment]
@auchri 
What is wrong with you?

News flash, to both of you. Not all humanity are coders or want to be coders or even should be coders. 

This program isn't made exclusively for coders.

Get over yourselves. 

We're allowed to make requests, and you're allowed to ignore them. 

At no point are you justified in scolding non-coders for being non-coders because that's a perfectly valid life choice and conversely we have Every right to make usability requests and even *gasp* complaints. Which you can again ignore.

If you don't want to hear complaints, keep all your code private and unsubscribe.

God, the entitlement. You guys apparently everyone to be like you, do all the work for you and kiss your ring when you volunteer to do some on your own and on top of all that you have to be snide and abusive on top of it.

Well, you don't see any of these people scolding you for all the crap they can do that you can't.

It's called diversity, look into it.

No one likes a bully. Keep that crap to yourself.

And remember you acted like this the next time you complain about anyone else's work in any context that you can't personally duplicate.

In short, grow up and get over yourselves. If you want your butt kissed, go tip heavy somewhere in the service industry.

**Comment by ThiagoLopes** [comment]
please, +1
@comzeradd -> +1

**Comment by dvborzov** [comment]
+1 add add spellchecking plz

**Comment by Zignd** [comment]
LOL it have been more than a year and this issue is still open.

**Comment by wolfjlupus** [comment]
+1 Plase sav su frmo ym mis-spelings! ;)

I wish we could donate money for features to be added ( if we can, tell me/us how ). I would give a few bucks to help support Telegram to add this to it's clients.

**Comment by shawnshaddock** [comment]
+1

**Comment by haevalencia** [comment]
+1
There are many reasons why it would be a good idea.

**Comment by anaptfox** [comment]
+1

**Comment by chmercesmoreira** [comment]
+1

**Comment by vitorgrs** [comment]
Please, would be good if the Windows version used the API (8+) https://msdn.microsoft.com/en-us/library/windows/desktop/hh869852(v=vs.85).aspx cause then it supports the touch keyboard for tablets too.

**Comment by pyed** [comment]
+1

**Comment by Placid7** [comment]
Guys Telegram is proprietary encrypted messaging. If/when spell check is incorporated it will be proprietary as well.

The very ability for you system to be able to access Telegram, (at the level spellcheck works) would completely undermine this security.

The obvious work around is using another app/program to write your messages and then send via Telegram.

If I'm wrong correct me.

rthomas529@gmail.com

**Comment by kmare** [comment]
@Placid7 you're wrong actually. The protocol is open source, all the apps acting as clients are open source etc (so they're not proprietary, and one of the many reasons I chose telegram in the first place). Implementing a spell checker has nothing to do with proprietary or security. In fact, it shouldn't even be hard to add that feature. I would do it myself if I had the time.

**Comment by Placid7** [comment]
Then I'm going to give a +1 @jacobmischka. My mind is blown that so many people are making this request and virtually no pull requests.

If it is open source and the protocol is available why isn't there a plugin being worked out by the users (which is what I cam here looking for)? I'm new to programming and don't have the skill to develope this from scratch and was actually looking for someone who was already working on this so I could take a wack at it.

**Comment by Zignd** [comment]
Maybe we should start a Kickstarter campaign in order to get some money to hire a C++ dev to implement this issue.

**Comment by jacobmischka** [comment]
Oh, I just read through [CONTRIBUTING.md](https://github.com/telegramdesktop/tdesktop/blob/master/CONTRIBUTING.md), if someone implemented this would it even be merged? @auchri

**Comment by dianjuar** [comment]
I will start to develop this feature. If some want to contribute, put in contact with me.

**Comment by stek29** [comment]
I'm sure that it child use system API's on OSX and Windows, and on Linux too, if any.
Don't invent a bicycle.

**Comment by jacobmischka** [comment]
Hunspell + Qt is common, certainly he wasn't planning on writing a spellchecking library from scratch. The system APIs make it a bit obnoxious making it cross platform, which I imagine will be the most annoying part.

**Comment by stek29** [comment]
@auchri Could you please check if Hunspell license is compatible with TDesktop one?

**Comment by auchri** [comment]
@stek29 Hunspell is triple licensed (`Free software (LGPL, GPL, MPL tri-license)`). Since [LGPL 2.1](https://github.com/hunspell/hunspell/blob/master/license.hunspell) [is compatible](http://www.gnu.org/licenses/license-list.en.html#LGPLv2.1) with GPLv3 (the license which Telegram Desktop [uses](https://github.com/telegramdesktop/tdesktop/blob/master/LICENSE)), it's save to use with tdesktop :)

**Comment by neesti** [comment]
Hello.
Please support Persian (Farsi) spell checker.
Thank you.

**Comment by slimshadie** [comment]
+1 for spell checking

**Comment by opeRaptor** [comment]
Wait... Opened in 2014 and still no solution? I really want this feature!

**Comment by Frederick888** [comment]
I think you can provide spellchecking for English first...

**Comment by perk11** [comment]
+1. Ever since we switched from Skype, spell checking is missed.

**Comment by AndydeCleyre** [comment]
I agree with @sandsmark's comment from 2 years ago, sonnet seems like a great choice here. Stick with Qt, stick cross platform.

https://github.com/KDE/sonnet
https://inqlude.org/libraries/sonnet.html

> Sonnet is a plugin-based spell checking library for Qt-based applications. It supports several different plugins, including HSpell, Enchant, ASpell and HUNSPELL.

> It also supports automated language detection, based on a combination of different algorithms.

> The simplest way to use Sonnet in your application is to use the SpellCheckDecorator class on your QTextEdit.

As long as you're using QTextEdit widgets, this would seem to be a very simple introduction.

I'd wager this project would see a healthy bump in pull requests if they moved to upstream Qt rather than a custom patched version (and consequently featured a simpler build process).

**Comment by Emadpres** [comment]
In FAQ you say "No Call for now. **We specialize in sending messages**, ... There are still many things that we can revolutionize in messaging".
I think spell checker is one of the _most-basic feature_ in every messaging app. Please add English spell checker for Windows client.

**Comment by Mte90** [comment]
we need that feature! on linux is very easy an integration of hunspell or other solutions

**Comment by zaps166** [comment]
If nobody else hasn't started working at hunspell implementation I can try to do this, but I don't know telegram code, so it can take a time... I've got simple code to Qt5 spell checker (via hunspell) - now it needs to be implemented in telegram-desktop.

**Comment by AndydeCleyre** [comment]
@telegramdesktop are you using a QTextEdit widget for the message composer box?

**Comment by zaps166** [comment]
@AndydeCleyre yes, this is QTextEdit. I haven't tried Sonnet, but it should work.

My hunspell highlighter works properly with Polish and English words. The most difficult thing is to create options for it :smile:

Screenshot (Telegram Desktop v0.9.56; Qt 5.6.1):
![telegram-spellchecker](https://cloud.githubusercontent.com/assets/6239090/17072666/51b2caf2-506a-11e6-8f4f-96da4907cf99.png)

---

> Unfortunately we do not merge any pull requests that have new feature implementations, translations to new languages and those which introduce any new user interface elements.

Hmm... What with this?

**Comment by zaps166** [comment]
I've got this: https://github.com/zaps166/tdesktop/commit/d4ed33144d35915d100eb27e574c89473e9b689d

Also attached here (forked repository can be changed or removed): [Add spell checker via hunspell.patch.txt](https://github.com/telegramdesktop/tdesktop/files/379486/Add.spell.checker.via.hunspell.patch.txt)

For me (and probably for users which want to compile tdesktop) it is sufficient :) The code uses system language as default spell checker and also always en_US as a second language (you can change it in code).
The patch can be applied also for stable (not master) version (just replace tabs by four spaces for `*.pro` file in the patch).
Also I recommend to use Qt 5.6.1 (only Windows code has rejects which can be ignored for Linux users) :smile:

For static linking (also for Windows and OS X) the hunspell probably must be patched to use dictionaries from Qt resources.
If the PRs with new features won't be merged (according to contribution guide) currently I won't finish it - developers can use this code or guide me what to do (add and fix - travis, code, \*.pro files; replace hard-coded paths and langs, etc.) for adding a new feature.

**Comment by perk11** [comment]
Can confirm, @zaps166's works wonderful.

**Comment by zaps166** [comment]
@auchri @john-preston Is it any chance for accepting PR with a new feature? If yes I can do some fixes and create a PR... Or just use my code (it is already signed in my repo) :)

**Comment by stek29** [comment]
@zaps166 did you test it on all platforms? Had you rewrote docs?
What about languages? What about non-default languages?
What about disabling spellcheck inside code blocks? (<pre>``` ... ```</pre> and <pre>` ... `</pre>)
What about use OS X's spellchecker on OS X, not hunspell's default one?

Btw why you haven't used kde's sonnet? I've thought it would be easier...

**Comment by zaps166** [comment]
@stek29

> did you test it on all platforms? Had you rewrote docs?

No, currently this is Linux-only (see `*.pro` file diff). Patch is not finished.

> What about languages? What about non-default languages?

It needs settings for this, as I wrote before - patch is not finished. Currently uses english and system language (if exists). Adding GUI for this with translations for supported languages by Telegram is more work than implementing spell-checking itself. If PR with new features are not accepted, so no need for doing this now (CONTRIBUTING.md).

> What about disabling spellcheck inside code blocks? (``` ... ``` and ` ... `)

Should be implemented, but currently it won't underline brackets and other characters, only variables and maybe some keywords.

> What about using OS X's spellchecker on OS X, not hunspell's default one?

I don't know Obj-C and OS X API. Hunspell is portable, so it should work there. Also Hunspell can unify spell-checking. Native spell-checking can be implemented anytime later.

> Have you checked compatibility with mentions by name (Blue text)?

Which blue text? I don't fully understand this sentence. Text with "@" prefix (it is blue)?

---

> Btw why you haven't used kde's sonnet? I've thought it would be easier...

Maybe I'll try later if it is easier (but Hunspell is already implemented...) - currently it is sufficient for me :smile:

**Comment by stek29** [comment]
@zaps166 blue text: try mentioning someone who has no username.

**Comment by zaps166** [comment]
@stek29 If someones name won't be in dictionary then it will underline it, but it still will be blue (currently).

**Comment by akashnimare** [comment]
@zaps166 @stek29 any updates on this?

**Comment by Pub4Game** [comment]
What? 2 years? @auchri just don it, omg

**Comment by auchri** [comment]
@Pub4Game I've got other things to do.

**Comment by Pub4Game** [comment]
@auchri I understand, but 2 years? It is difficult to add the change event of the field and to add a spell check?

**Comment by stek29** [comment]
@Pub4Game If it's so easy why won't you do it?

**Comment by Innomen** [comment]
@stek29

I'm so tired of that "logic."

What's easy for a cardiac surgeon is not easy for a physicist.

Just because someone can recognize something is easy for someone else doesn't mean it's easy for them.

People aren't ants. We have individual variation.

Coders these days, infected by greed almost to a man, don't take a dump without a royalty fee in the offing, regardless of ease. When they aren't simply lying about being coders in the first place.

Heh, but what do we expect from the only industry on earth where being called a "hack" isn't an insult?

**Comment by john-preston** [comment]
@Innomen It's really not that easy for me, sorry :(

**Comment by Innomen** [comment]
@john-preston

That's ok :) I suspect if you could, you would, out of simple kindness, which is a lot more than I can say for some people.

**Comment by john-preston** [comment]
@Innomen Btw maybe you at least know where to get a set of dictionaries for the spellchecking, that are completely license-free, can be embedded into tdesktop app and are available for at least all the supported tdesktop languages? (English, Spanish, Italian, Dutch, Korean, German, Portuguese-Brazil)

**Comment by Innomen** [comment]
@john-preston

There are none for english afaik. Everyone that is capable of compiling one is apparently either too narcissistic or too greedy to do it for the good of the art/species. Or, is prevented from doing so by laws designed by and for said greedy narcissists, all of whom apparently want to patent and license the vowels when they grow up.

**Comment by AndydeCleyre** [comment]
What's wrong with Sonnet and its SpellCheckDecorator here? It supports HSpell, Enchant, ASpell and HUNSPELL, at least.

**Comment by AndydeCleyre** [comment]
I know for sure there are hunspell dictionaries for English, Spanish, Italian, Dutch, Korean, German, Portuguese-Brazil, and I think they're all LGPL/scowl. Not sure about korean dicts.

**Comment by nop144d** [comment]
@AndydeCleyre https://inqlude.org/libraries/sonnet.html says Platform:Linux, is it cross-platform actually?

**Comment by auchri** [comment]
@nop144d 

https://api.kde.org/frameworks/sonnet/html/index.html says:

> Supported platforms
> Linux, MacOSX, Windows

**Comment by nop144d** [comment]
@auchri ah, ok. So what's the problem with this library then, if any?

**Comment by auchri** [comment]
@nop144d I don't know why nobody made a PR ;)

**Comment by AndydeCleyre** [comment]
@auchri not all of us are C++ devs, and as is well known this build process is painful.

**Comment by lordkage** [comment]
@auchri Didn't @zaps166 on this [comment](https://github.com/telegramdesktop/tdesktop/issues/96#issuecomment-234710869) and this [comment](https://github.com/telegramdesktop/tdesktop/issues/96#issuecomment-236019723) ask about what else was needed for the PR to be excepted?

**Comment by jcjolley** [comment]
I would love spell check support, even if it didn't provide suggestions, just provided the red squiggly under misspelled words.

**Comment by joonaskokko** [comment]
+1 to getting the native OSX spellchecker to work.

**Comment by TrendaFet** [comment]
Agreed. Even an indicator for a misspell would be greatly appreciated, I never knew how much I needed one until I realized that without it, I type like a monkey. 

You have no idea how many corrections I had to make in this comment.

**Comment by Innomen** [comment]
I feel the need to clarify that indication of a misspelling is also sufficient in my case. Assuming that simplicity speeds along implementation of course.

**Comment by zaps166** [comment]
[Telegram Desktop with Hunspell](https://github.com/zaps166/tdesktop-hunspell) ([patch](https://github.com/zaps166/tdesktop-hunspell/blob/master/Hunspell.diff) + [Arch Linux package](https://github.com/zaps166/tdesktop-hunspell/releases)).

**Comment by stormyqq** [comment]
I would like it too. without spelling checking the telegram app is useless to me, as I am not a native english speaker. so it's important for me to know whether what I am typing is correct or wrong.

**Comment by dianjuar** [comment]
@auchri I tried once. But in that time has terrible painful set up the development environment, the doc has based on debian and I have Arch Linux. 
Maybe create a docker will be the best practice.

**Comment by Sbzz** [comment]
Telegram Desktop with Hunspell (patch + Arch Linux package).  - And how do you install this on say an Ubuntu machine? Thanks.

**Comment by stek29** [comment]
@Sbzz You can install Arch on it :>

**Comment by Sbzz** [comment]
@stek29  This is needed on an Ubuntu machine. Any solution?

**Comment by stek29** [comment]
@Sbzz You'd have to build it on your own.

**Comment by Sbzz** [comment]
@stek29 In order words, no public solution as yet.

**Comment by zaps166** [comment]
@Sbzz If you have new glibc you can try to copy all missing/obsolete dependencies from Arch Linux to unpacked Telegram directory and set `LD_LIBRARY_PATH`.

**Comment by akashnimare** [comment]
I have built an unofficial telegram desktop client using electron and it supports spellchecker as well. 

Check it out -

https://github.com/akashnimare/telectron

![spell](https://cloud.githubusercontent.com/assets/2263909/24840730/65bea040-1d90-11e7-9f75-376cfa512849.jpg)

**Comment by Innomen** [comment]
OOOOoooooo /hyped

**Comment by kirsan31** [comment]
https://github.com/telegramdesktop/tdesktop/pull/3232

**Comment by stek29** [comment]
@akashnimare but can't I just use any browser which has exact same spell checking, but a lot more features (like receiving notifications when tab with webogram is closed -- I can't see your app sending notifications when it's closed).

**Comment by AdKiller** [comment]
Please make it happen. It shouldn't be that hard considering there are ready-made spell checking solutions out there...

**Comment by stormyqq** [comment]
want for windows if it's possible. thanks.

**Comment by Mte90** [comment]
any updates on that? In Mozilla we use telegram a lot and will be helpful to avoid so many typos!

**Comment by lordkage** [comment]
@Mte90 There have been several people that have created PRs, asked for clarifications, and volunteered for to assist. And none of it has gotten a response from the project owner. Has Mozilla created created a fork with a working release for all platforms?

**Comment by stek29** [comment]
@lordkage I've seen just one PR, and that PR got a lot of attention and help from John.

**Comment by lordkage** [comment]
That is awesome news!
Does that mean we are close to having resolution?

**Comment by AdKiller** [comment]
so they just implemented the extremely important "Autocompletion for emoji" and still no spell checker... sad

**Comment by gsigler** [comment]
+1 telegram desktop needs this feature..

**Comment by gabrielpaim** [comment]
This feature would be indeed awesome for Telegram! I'm really hoping to see spell checking soon!

**Comment by mtleitner** [comment]
Spell checking on Telegram would be great !!!

**Comment by jcjolley** [comment]
3 years later, and still nothing.  :(

**Comment by Innomen** [comment]
/type critical thing about oligarch owned insecure closed source server chat app...

"auchri deleted a comment from" in 3... 2... 1...

Guys, seriously, never gonna happen. I'm unsubscribing now, you should do the same and try to migrate to a fully open source solution.

**Comment by john-preston** [comment]
@Innomen Please stop posting offtopic here, this issue has nothing to do with server side source code, project ownership etc.

This is completely client-side issue and it can be done by the community with source of this app being open. So it will indeed be required to delete your offtopic comments :(

@auchri helps with issue management in his free time, he has no interest in this apart from the improvement of the application. The deleted comments were not critical, they just were meaningless, because they could be easier done by reactions.

**Comment by Innomen** [comment]
@john-preston It's not off topic. It very likely explains why there's no movement on both trivial (like this) or critical issues, like log exporting and deletion. (Let alone server code we'll clearly never see.)

Speaking of which, calling this open source is like calling windows 10 open source because I can write and share a shell for it.

https://en.wikipedia.org/wiki/Telegram_(messaging_service)#Servers

You need to free yourself from throwing good money after bad. You should cut your losses, and move to a project that isn't feeding logs straight to the global Intelligence community.

Auchri codes for this project to maintain his status and ego. It's like you've never met a moderator/admin/coder before. 

But go on, keep trying to tell us he's doctors without borders.

Muting this thread again. 

/disgusted

**Comment by 23rd** [comment]
It's not over yet. The spellchecker for Linux and Windows 7 is in the WIP state.
